paymentsheet-example/src/main/java/com/stripe/android/paymentsheet/example/samples/ui/server_side_confirm/ServerSideConfirmationViewModel.kt

```diff
@@ -0,0 +1,251 @@
+                val errorMessage = "Unable to create intent\n${apiResult.error.exception}"
+                CreateIntentCallback.Result.Failure(
+                    cause = RuntimeException(errorMessage),
+                    displayMessage = "Something went wrong…",
+                )
+            }
+        }
+    }
+
+    fun handlePaymentSheetResult(paymentResult: PaymentSheetResult) {
+        val status = when (paymentResult) {
+            is PaymentSheetResult.Canceled -> null
+            is PaymentSheetResult.Completed -> "Success"
+            is PaymentSheetResult.Failed -> paymentResult.error.message
+        }
+
+        _state.update {
+            it.copy(
+                isProcessing = false,
+                status = status,
+                didComplete = paymentResult is PaymentSheetResult.Completed,
+            )
+        }
+    }
+
+    fun updateQuantity(productId: CartProduct.Id, newQuantity: Int) {
+        val dirtyCartState = state.value.cartState.updateQuantity(productId, newQuantity)
+        _state.update {
+            it.copy(dirtyCartState = dirtyCartState)
+        }
+    }
+
+    fun updateSubscription(isSubscription: Boolean) {
+        val dirtyCartState = state.value.cartState.copy(
+            isSubscription = isSubscription,
+            requiresCalculation = true,
+        )
+        _state.update {
+            it.copy(dirtyCartState = dirtyCartState)
+        }
+    }
+
+    fun handleBuyButtonPressed() {
+        _state.update {
+            it.copy(isProcessing = true)
+        }
+    }
+
+    private suspend fun prepareCheckout() {
+        val currentState = _state.updateAndGet {
+            it.copy(isProcessing = true)
+        }
+
+        val request = currentState.cartState.toCheckoutRequest()
+        val requestBody = Gson().toJson(request)
+
+        val apiResult = Fuel
+            .post("$backendUrl/checkout")
+            .jsonBody(requestBody)
+            .suspendable()
+            .awaitModel<ExampleCheckoutResponse>()
+
+        when (apiResult) {
+            is ApiResult.Success -> {
+                val newCartState = currentState.cartState.updateWithResponse(apiResult.value)
+
+                PaymentConfiguration.init(
+                    context = getApplication(),
+                    publishableKey = apiResult.value.publishableKey,
+                )
+
+                _state.update {
+                    it.copy(
+                        isProcessing = false,
+                        isError = false,
+                        confirmedCartState = newCartState,
+                        dirtyCartState = null,
+                        requiresFlowControllerConfigure = true,
+                    )
+                }
+            }
+            is ApiResult.Failure -> {
+                val status = "Preparing checkout failed\n${apiResult.error.message}"
+                _state.update {
+                    it.copy(isProcessing = false, isError = true, status = status)
```

Comment:
Do we provide a retry mechanism?

Reply:
Not explicitly. A merchant could simply call `configureWithIntentConfiguration` again, or they reset the state and show an error message (which is what we’re doing here).

Reply:
I'm more talking about a user retry mechanism as part of our example.

I assume the screen fails here, and the only option is to go back, and re-launch the activity.

Reply:
That’s right. For the `prepareCheckout` call, we’d currently leave the screen in an unusable state. I added an error state with a retry button.